File: Python/Binary Search Tree/Inorder_Successor_BST_I.py
# Inorder Successor given a root of BST 
# Author: Pavan Kumar Paluri
# Leetcode Question: https://leetcode.com/problems/inorder-successor-in-bst/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode':
        def right_successor(node):
            node = node.right
            while node.left is not None:
                node = node.left
            return node
        if root is None:
            return None
        if p.right is not None:
            return right_successor(p)
        else:
            list_node = []
            def helper(root, p):
                # inorder traversal
                # inorder: [left, root, right]
                if root is None:
                    return None
                helper(root.left, p)
                if root.val > p.val:
                    list_node.append(root)
                    return root
                helper(root.right, p)
            helper(root, p)
            if len(list_node) >0:
                return list_node[0]
            else:
                return None

File: C++/Binary Search Tree/Inorder_Successor_BST_I.cpp
// Inorder Successor Binary-Search Tree
// Using Iteration + Inorder Traversal + Node Stack
// Author: Pavan Kumar Paluri

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* right_successor(TreeNode* node)
    {
        // To find the immediate right successor, go one node to root and then left
        node = node->right;
        while(node->left != nullptr)
        {
            node = node->left;
        }
        // if here: node is the leftmost-leaf node
        return node;
    }
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        // if p has right node:
        if(p->right != nullptr)
            return right_successor(p);
        // if p does not have right node:
        vector<TreeNode*> node_stack;
        // begin from root node
        if(root == nullptr)
            return nullptr;
        int prev_node_val = INT_MIN;
        // Go to the leftmost node
        while(node_stack.size()>0 || root != nullptr)
        {
            while(root!=nullptr)
            {
                node_stack.push_back(root);
                root = root->left;
            }
            // last-leaf leftmost node: root
            root = node_stack[node_stack.size()-1];
            node_stack.pop_back();
            if(prev_node_val == p->val)
            {
                return root;
            }
            prev_node_val = root->val;
            root = root->right;
            
        }
        return nullptr;
        }
};
